services/config.go

```diff
@@ -0,0 +1,241 @@
+package services
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/getlantern/detour"
+	"github.com/getlantern/golog"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/getlantern/flashlight/v7/apipb"
+	"github.com/getlantern/flashlight/v7/common"
+	"github.com/getlantern/flashlight/v7/ops"
+)
+
+const (
+	defaultConfigSaveDir  = ""
+	defaultConfigFilename = "proxies.conf"
+
+	defaultConfigPollInterval = 3 * time.Minute
+	defaultConfigPollJitter   = 2 * time.Minute
+)
+
+// ConfigOptions specifies the options to use for ConfigService.
+type ConfigOptions struct {
+	// URL to use for fetching this config.
+	OriginURL string
+
+	// UserConfig contains data for communicating the user details to upstream
+	// servers in HTTP headers, such as the pro token and other options.
+	UserConfig common.UserConfig
+
+	// RoundTripper provides the http.RoundTripper the fetcher should use, which allows us to
+	// dictate whether the fetcher will use dual fetching (from fronted and chained URLs) or not.
+	RoundTripper http.RoundTripper
+
+	// PollInterval specifies how frequently to poll for new config.
+	PollInterval time.Duration
+	// PollJitter specifies the max amount of jitter to add to the poll interval.
+	PollJitter time.Duration
+}
+
+type configService struct {
+	opts          *ConfigOptions
+	clientInfo    *apipb.ConfigRequest_ClientInfo
+	configHandler ConfigHandler
+	lastFetched   time.Time
+
+	sender *sender
+
+	done    chan struct{}
+	running bool
+	// runningMu is used to protect the running field.
+	runningMu sync.Mutex
+}
+
+// ConfigHandler handles updating and retrieving the client config.
+type ConfigHandler interface {
+	// GetConfig returns the current client config.
+	GetConfig() *apipb.ConfigResponse
+	// SetConfig sets the client config to the given config.
+	SetConfig(new *apipb.ConfigResponse)
+}
+
+var _configService = &configService{sender: &sender{}}
```

Comment:
It looks like all references from `_configService` are only contained at the `StartConfigService`, can we move it to inside the function and I guess we could call `cfgService` or just `service`

Reply:
Otherwise this variable will be initialized when the application is starting or loading the services package, if this is the intention it's ok but I guess it isn't since isn't referred outside the package

Reply:
This so `_configService` is a singleton to prevent multiple instances of it being started. 

> Otherwise this variable will be initialized when the application is starting or loading the services package, if this is the intention it's ok but I guess it isn't since isn't referred outside the package

This is fine since `StartConfigService` is going to be called during startup. I intentionally initialized it so there wouldn't have to be a mutex that would have to be locked/unlocked just to check if it was null and then initialized it.